Allow AWS Assume Role

ekristen opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.
-->

What does this PR change? What problem does it solve?
-----------------------------------------------------

- Adds support for Assuming a Role in AWS

Resolves #4472 

It adds the ability to allow the discovered credentials for AWS to assume another role and use those credentials for the actual run of restic.

Example: Personal IAM user (ekristen) has access to ec2 console but not s3 backup location but has the right to assume the role, backup, backup has full access to the s3 bucket, so this allows restic to assume a role.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Not really, opened #4472

<!--
Link issues and relevant forum posts here.

If this PR resolves an issue on GitHub, use "Closes #1234" so that the issue
is closed automatically when this PR is merged.
-->

Testing this Pull Request
-------------------------
Using the files from this gist create the folllowing, a test iam user, test iam role, and test s3 bucket. https://gist.github.com/ekristen/e20a976653d0624957add5b4d2ef64ba

1. Create a set of IAM creds for the test user and export them using AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
2. Export AWS_ASSUME_ROLE_ARN with the value of your test iam role arn
3. Run restic against your newly created s3 bucket. 

*Note:** for more advanced use cases you can pass AWS_ASSUME_ROLE_SESSION_NAME and AWS_ASSUME_ROLE_EXTERNAL_ID and use conditions to trust based on values.  Additionally you can pass AWS_ASSUME_ROLE_POLICY which is a JSON document for an IAM policy, this can be used to set a STS policy to restrict the STS credentials to a specific path as an example.

Checklist
---------

<!--
You do not need to check all the boxes below all at once. Feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box. Enable a checkbox by replacing [ ] with [x].
-->

Note: tests -- there aren't extensive existing tests on authentication methods, there's not a framework to build off of that I could see to add tests.

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

ekristen committed:
feat: support AWS assume role

ekristen commented:
This is true. I wonder how often backups exceed 12 hours and use s3 and would use assume role? 

Creating a go func that goes into a loop at expiration/2 and updates the current credentials would work. However back to my original question of how often the exceeding 12 hours would happen.

I see two options

1. Update docs for initial implementation to state that there's a limit of 12 hours with role assumption and cannot use if backups exceed.
2. Work on a for loop to renew credentials at half the expiration time to allow further calls to continue to work.

MichaelEischer reviewed (changes requested):
Using the aws-sdk in addition to minio is not an option.

From a quick look at the minio-go documentation, it also supports assuming a role: https://pkg.go.dev/github.com/minio/minio-go/v7@v7.0.61/pkg/credentials#STSAssumeRole . Please use that approach instead. That will likely also just fix the expiry problems.

ekristen commented:
@MichaelEischer Why is it a no-go, it's a small part of the library? The minio doesn't support all the necessary options like passing an ExternalId. I can write a wrapper to support the expiration. 

Other than the above, is the rest of it acceptable? The environment variable names etc?

MichaelEischer commented:
> Why is it a no-go, it's a small part of the library?

It increases the binary size by 1MB, seems to work for AWS only, feels completely out of place compared to the existing credentials methods and so on.

>  The minio doesn't support all the necessary options like passing an ExternalId. I can write a wrapper to support the expiration.

I'd prefer to ask whether the minio developers are willing to add support for ExternalId to minio-go . There are already two AWS specific options for STSAssumeRole, That would be a 10-20 lines change on the minio side along with proper integration in the library, compared to the brittle 100+ lines implementation here.

ekristen commented:
Gotcha. I will make the necessary changes and pull requests.

ekristen commented:
@MichaelEischer changes as requested, however it depends upon changes to minio-go/v7 which I've opened a PR for here https://github.com/minio/minio-go/pull/1887 -- I'm happy to leave my fork up if you'd like to merge as is, but if you'd like to wait and see if the PR can get merged mainstream, we can leave this open until that PR is merged and upgrade the dependency accordingly.

MichaelEischer commented on internal/backend/s3/s3.go:
```diff
@@ -87,6 +87,53 @@ func open(ctx context.Context, cfg Config, rt http.RoundTripper) (*Backend, erro
+			SecretKey: c.SecretAccessKey,
+		}
+		if c.SessionToken != "" {
+			opts.SessionToken = c.SessionToken
```
As these options default to empty strings, there's no need for the surrounding if. This also allows inlining `os.Getenv("AWS_ASSUME_ROLE_SESSION_NAME")`, external Id and policy.

ekristen replied:
Done.

MichaelEischer commented on internal/backend/s3/s3.go:
```diff
@@ -87,6 +87,53 @@ func open(ctx context.Context, cfg Config, rt http.RoundTripper) (*Backend, erro
+		if policy != "" {
+			opts.Policy = policy
+		}
+		if len(os.Getenv("AWS_REGION")) > 0 {
```
I think it would be more elegant to save `$AWS_REGION` to a variable once.

ekristen replied:
Done

MichaelEischer commented on internal/backend/s3/s3.go:
```diff
@@ -87,6 +87,53 @@ func open(ctx context.Context, cfg Config, rt http.RoundTripper) (*Backend, erro
 	}
 
+	// assume role
+	roleArn := os.Getenv("AWS_ASSUME_ROLE_ARN")
```
Please extract the code into a helper function. The `open()` function is already far too long.

ekristen replied:
Done

MichaelEischer reviewed (commented):
Thanks for updating the PR. Yes, I'd prefer to wait until the minio PR is merged.

ekristen commented:
@MichaelEischer the minio PR was just merged https://github.com/minio/minio-go/pull/1887#event-10625033676 -- do we need to wait until it hits a release? If not I can update the branch to use the specific commit.

ekristen commented:
Working on fully testing the changes right now, but for anyone else that wants to here's the docker image.
```
ghcr.io/ekristen/restic:0.16.0-ek2
```

ekristen commented:
Had to make a tweak to the region code to make things align properly with assuming roles properly.
```
ghcr.io/ekristen/restic:0.16.0-ek3
```

MichaelEischer commented:
> @MichaelEischer the minio PR was just merged [minio/minio-go#1887 (comment)](https://github.com/minio/minio-go/pull/1887#event-10625033676) -- do we need to wait until it hits a release? If not I can update the branch to use the specific commit.

I'd prefer to wait until the next version of the minio-go library is released, the next restic release will still take quite some time, so there's no need to rush it.

I'll try to have a closer look at the code in the next days/weeks.

ekristen commented:
@MichaelEischer understood.

ekristen commented:
@MichaelEischer thinking about this, I think we need to at least prefix these with `RESTIC_` to avoid conflicts in the future.

ekristen commented:
@MichaelEischer minio-go release a new version, this includes that update to make this PR work now.

I've also updated the env vars to be prefixed with `RESTIC_` for the AWS side to prevent any possible collisions with other tooling.

Here's a build docker image for anyone that wants to test.

```
ghcr.io/ekristen/restic:0.16.0-ek5
```

MichaelEischer commented:
> @MichaelEischer minio-go release a new version, this includes that update to make this PR work now.

Thanks, as a heads-up I'm pretty busy at the moment, I probably won't be able to review the PR before Christmas, sorry.

MichaelEischer committed:
s3: check for EnvAWS credentials before Static credentials

EnvAWS considers more environment variables, including AWS_SESSION_TOKEN
and thus should be checked first.

MichaelEischer committed:
s3: minor code cleanups

MichaelEischer committed:
s3: update documentation

MichaelEischer commented on internal/backend/s3/s3.go:
```diff
@@ -107,13 +107,13 @@ func getCredentials(cfg Config) (*credentials.Credentials, error) {
 	//    call to a pre-defined endpoint, only valid inside
 	//    configured ec2 instances)
 	creds := credentials.NewChainCredentials([]credentials.Provider{
+		&credentials.EnvAWS{},
```
I've moved the `&credentials.EnvAWS{}` credentials back to the top of the list, as I'm pretty sure that it makes a difference. Only `EnvAWS` handles the `AWS_SESSION_TOKEN` environment variable.

ekristen replied:
It could cause problems in my opinion.

You are correct that it handles AWS_SESSION_TOKEN, but static config should supersede all.

If you have env vars set and static config set, with your change the env vars will always take precedence which I feel is unexpected behavior.

So if static is first, if someone has passed in static creds it'll work as it does not require a session token.

Otherwise, if a session token is required then environment variables or an AWS profile is the only way to go.

MichaelEischer replied:
> It could cause problems in my opinion.

Well, before this PR `EnvAWS` was the first entry, thus keeping it at the position won't cause (new) problems.

In addition, the static config variables behave different than you expect: `cfg.KeyID` and `cfg.Secret` are filled from the environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`. An `AWS_SESSION_TOKEN` is always used together with a key_id and access_key, such that this combination would break if `credentials.Static` is moved before `EnvAWS`.

So let's just keep the order as it was before this PR.

ekristen replied:
I see. That should probably not be the case, but I understand not changing it.

Just know that if the environment variables are set, it will always just use EnvAWS, the only time it'll use static is if the `cfg.KeyID` and `cfg.Secret` are explicitly set without environment variables.

MichaelEischer commented on doc/040_backup.rst:
```diff
@@ -628,6 +628,12 @@ environment variables. The following lists these environment variables:
+    RESTIC_AWS_ASSUME_ROLE_ARN          Amazon IAM Role ARN to assume using discovered credentials
+    RESTIC_AWS_ASSUME_ROLE_SESSION_NAME Session Name to use with the role assumption
+    RESTIC_AWS_ASSUME_ROLE_EXTERNAL_ID  External ID to use with the role assumption
+    RESTIC_AWS_ASSUME_ROLE_POLICY       Inline Amazion IAM session policy
```
I've tried to come up with a description for `RESTIC_AWS_ASSUME_ROLE_POLICY` and `RESTIC_AWS_ASSUME_ROLE_STS_ENDPOINT` which were missing. Please take a look.

ekristen replied:
I think the policy description is acceptable. If someone wants to use it they'll need to understand how STS policies work anyways and look up AWS documentation.

MichaelEischer reviewed (approved):
Sorry for the long delay! I've squashed the commits (rebasing all the changes would have been too messy) and rebased them to the current master branch. That implicitly updates minio to v7.0.66. 

I've added two remarks, please take a look. Other than that I'm happy with the changes.

Could you give the PR one final try to check that everything still works?

ekristen commented on doc/040_backup.rst:
```diff
@@ -628,6 +628,12 @@ environment variables. The following lists these environment variables:
+    RESTIC_AWS_ASSUME_ROLE_EXTERNAL_ID  External ID to use with the role assumption
+    RESTIC_AWS_ASSUME_ROLE_POLICY       Inline Amazion IAM session policy
+    RESTIC_AWS_ASSUME_ROLE_REGION       Region to use for IAM calls for the role assumption (default: us-east-1)
+    RESTIC_AWS_ASSUME_ROLE_STS_ENDPOINT URL to the STS endpoint (default is determined based on RESTIC_AWS_ASSUME_ROLE_REGION)
```
```suggestion
    RESTIC_AWS_ASSUME_ROLE_STS_ENDPOINT URL to the STS endpoint (default is determined based on RESTIC_AWS_ASSUME_ROLE_REGION) You generally do not need to set this, advanced use only.
```

MichaelEischer replied:
I've updated the description.

ekristen commented:
Hey @MichaelEischer all good, glad you found time to come back around to this! I made a couple comments on your comments.

I'll go build and test this out and reply once done.

ekristen commented:
Everything seems to be working for me just fine.

MichaelEischer commented:
Thanks for testing, then let's merge it!
